source/common/http/utility.cc

```diff
@@ -749,48 +751,153 @@ std::string Utility::PercentEncoding::decode(absl::string_view encoded) {
-  // effort attempt.
-  const auto colon_pos = host.rfind(':');
-  absl::string_view host_to_resolve = host;
-  absl::optional<uint16_t> port;
-  if (colon_pos != absl::string_view::npos && host_to_resolve.back() != ']') {
-    const absl::string_view string_view_host = host;
-    host_to_resolve = string_view_host.substr(0, colon_pos);
-    const auto port_str = string_view_host.substr(colon_pos + 1);
-    uint64_t port64;
-    if (port_str.empty() || !absl::SimpleAtoi(port_str, &port64) || port64 > 65535) {
-      // Just attempt to resolve whatever we were given. This will very likely fail.
-      host_to_resolve = host;
+class MalformedAuthority {
+  static void throwException(absl::string_view authority) {
+    throw EnvoyException(fmt::format("malformed authority: {}", authority));
+  }
+
+  friend absl::string_view Utility::hostFromAuthority(absl::string_view authority);
+};
+
+absl::string_view Utility::hostFromAuthority(absl::string_view authority) {
+  bool ipv6_flag = false;
+
+  if (authority.find("[") != absl::string_view::npos) {
+    ipv6_flag = true;
+  }
+
+  auto bracket_close_pos = authority.rfind("]");
+  if (ipv6_flag) {
+    if (bracket_close_pos != absl::string_view::npos) {
+      return authority.substr(1, bracket_close_pos - 1);
     } else {
-      port = static_cast<uint16_t>(port64);
+      MalformedAuthority::throwException(authority);
     }
   }
 
-  // Now see if this is an IP address. We need to know this because some things (such as setting
-  // SNI) are special cased if this is an IP address. Either way, we still go through the normal
-  // resolver flow. We could short-circuit the DNS resolver in this case, but the extra code to do
-  // so is not worth it since the DNS resolver should handle it for us.
-  bool is_ip_address = false;
-  try {
-    absl::string_view potential_ip_address = host_to_resolve;
-    // TODO(mattklein123): Optimally we would support bracket parsing in parseInternetAddress(),
-    // but we still need to trim the brackets to send the IPv6 address into the DNS resolver. For
-    // now, just do all the trimming here, but in the future we should consider whether we can
-    // have unified [] handling as low as possible in the stack.
-    if (potential_ip_address.front() == '[' && potential_ip_address.back() == ']') {
-      potential_ip_address.remove_prefix(1);
-      potential_ip_address.remove_suffix(1);
+  auto comma_pos = authority.rfind(":");
+  if (comma_pos != absl::string_view::npos) {
+    return authority.substr(0, comma_pos);
+  } else {
+    return authority;
+  }
+}
+
+absl::optional<absl::string_view> Utility::portFromAuthority(absl::string_view authority) {
+  const auto pos = authority.rfind(":");
+  if (pos == absl::string_view::npos) {
+    return absl::nullopt;
+  }
+  return authority.substr(pos + 1);
+}
+
+const Utility::AuthorityAttributes Utility::parseAuthority(absl::string_view authority) {
```

Comment:
This function together with hostFromAuthority, portFromAuthority, parseInternetAddress and parseInternetAddressAndPort does a lot of redundant scanning of the authority string. I think the detection and extraction of values should be combined as much as possible. Some points:

1.  Check if it IPv6 address by looking if the first character is [. If it is there is no need to check if it is an IPv4 address or DNS name.
1.  Limit scan for the trailing port number to 6 characters
1.  Check if the name shorter than a possible IPv4 address and does and feed it to the inet_pton. There is not need to detect if a string resembles an IPv4 address before as it just doubles the work.